File: kmp-native/shared/src/commonMain/kotlin/com/stackbenchmark/kmpnative/data/repository/MovieRepositoryImpl.kt
package com.stackbenchmark.kmpnative.data.repository

import com.stackbenchmark.kmpnative.data.mapper.toDomain
import com.stackbenchmark.kmpnative.data.remote.TmdbApiClient
import com.stackbenchmark.kmpnative.domain.entity.MovieDetail
import com.stackbenchmark.kmpnative.domain.entity.MoviePage
import com.stackbenchmark.kmpnative.domain.repository.MovieRepository

class MovieRepositoryImpl(
    private val apiClient: TmdbApiClient
) : MovieRepository {

    override suspend fun getPopularMovies(page: Int): Result<MoviePage> {
        return try {
            val response = apiClient.getPopularMovies(page)
            Result.success(response.toDomain())
        } catch (e: Exception) {
            Result.failure(e)
        }
    }

    override suspend fun getMovieDetail(id: Int): Result<MovieDetail> {
        return try {
            val response = apiClient.getMovieDetail(id)
            Result.success(response.toDomain())
        } catch (e: Exception) {
            Result.failure(e)
        }
    }
}

File: rn-expo/src/data/MovieRepositoryImpl.ts
import { MovieDetail, MoviePage } from "../domain/entities";
import { MovieRepository, Result } from "../domain/MovieRepository";
import {
  mapMovieDetailDtoToDomain,
  mapPopularMoviesResponseToDomain,
} from "./mappers";
import { TmdbApiClient } from "./TmdbApiClient";

export class MovieRepositoryImpl implements MovieRepository {
  constructor(private readonly apiClient: TmdbApiClient) {}

  async getPopularMovies(page: number): Promise<Result<MoviePage>> {
    try {
      const dto = await this.apiClient.getPopularMovies(page);
      return { success: true, data: mapPopularMoviesResponseToDomain(dto) };
    } catch (error) {
      const message =
        error instanceof Error ? error.message : "An unknown error occurred";
      return { success: false, error: message };
    }
  }

  async getMovieDetail(id: number): Promise<Result<MovieDetail>> {
    try {
      const dto = await this.apiClient.getMovieDetail(id);
      return { success: true, data: mapMovieDetailDtoToDomain(dto) };
    } catch (error) {
      const message =
        error instanceof Error ? error.message : "An unknown error occurred";
      return { success: false, error: message };
    }
  }
}
